cosmo_tester/test_suites/snapshots/__init__.py

```diff
@@ -62,58 +278,260 @@ def wait_for_execution(client, execution, logger):
+
+
+def upload_snapshot(manager, local_path, SNAPSHOT_ID, logger):
+    logger.info('Uploading snapshot to latest manager..')
+    snapshot = manager.client.snapshots.upload(local_path,
+                                               SNAPSHOT_ID)
+    logger.info('Uploaded snapshot:%s%s',
+                os.linesep,
+                json.dumps(snapshot, indent=2))
+
+
+def restore_snapshot(manager, SNAPSHOT_ID, cfy, logger):
+    # Show the snapshots, to aid troubleshooting on failures
+    manager.use()
+    cfy.snapshots.list()
+
+    logger.info('Restoring snapshot on latest manager..')
+    restore_execution = manager.client.snapshots.restore(
+        SNAPSHOT_ID,
+    )
+    try:
+        restore_execution = wait_for_execution(
+            manager,
+            restore_execution,
+            logger)
+    except ExecutionFailed:
+        # See any errors
+        cfy.executions.list(['--include-system-workflows'])
+        raise
+
+
+def check_plugins(manager, old_plugins, logger, tenant=None):
+    if tenant:
+        logger.info('Checking plugins for {tenant}'.format(tenant=tenant))
+    else:
+        logger.info('Checking plugins')
+    plugins = get_plugin_list(manager, tenant)
+    assert plugins == old_plugins
+
+    # Now make sure they're correctly installed
+    with manager.ssh() as fabric_ssh:
+        for plugin_name, plugin_version, _ in plugins:
+            path = INSTALLED_PLUGIN_PATH.format(
+                tenant=tenant or 'default_tenant',
+                name=plugin_name,
+                version=plugin_version,
+            )
+            logger.info('Checking plugin {name} is in {path}'.format(
+                name=plugin_name,
+                path=path,
+            ))
+            fabric_ssh.run('test -d {path}'.format(path=path))
+            logger.info('Plugin is correctly installed.')
+
+    if tenant:
+        logger.info('Plugins as expected for {tenant}'.format(tenant=tenant))
+    else:
+        logger.info('Plugins as expected.')
+
+
+def check_deployments(manager, old_deployments, logger, tenant=None):
+    deployments = get_deployments_list(manager, tenant)
+    assert deployments == old_deployments
+
+    if tenant:
+        logger.info('Checking deployments for {tenant}'.format(tenant=tenant))
+    else:
+        logger.info('Checking deployments.')
+    # Now make sure the envs were recreated
+    with manager.ssh() as fabric_ssh:
+        for deployment in manager.client.deployments.list():
+            path = DEPLOYMENT_ENVIRONMENT_PATH.format(
+                tenant='default_tenant',
+                name=deployment['id'],
+            )
+            logger.info(
+                'Checking deployment env for {name} was recreated.'.format(
+                    name=deployment['id'],
+                )
+            )
+            fabric_ssh.run('test -d {path}'.format(path=path))
+            logger.info('Deployment environment was recreated.')
+    if tenant:
```

Comment:
Ok, after enough of those, I think there's a place for a func like:

```
  def log_with_tenant(msg, log_type=INFO, tenant=None):
      if log_type==INFO:
          log = logger.info
      elif ...
    
      if tenant:
           msg += ' for tenant {0}'.format(tenant)

      log(msg)
```

If you disagree, that's fine, but consider it - there are _12_ `if tenant` statements here.